File: backend/app/routers/reports.py
from datetime import datetime, timezone
from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status

from app.auth import get_current_admin
from app.db import get_supabase, get_supabase_client
from app.schemas import ReportResponse, StatusHistoryEntry, StatusUpdateRequest
from app.services.storage import upload_report_image

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("", response_model=list[ReportResponse])
async def list_reports() -> list[dict]:
    supabase = get_supabase()
    response = supabase.table("reports").select("*").order("created_at", desc=True).execute()
    if response.data is None:
        return []
    return response.data


@router.post("", response_model=ReportResponse, status_code=status.HTTP_201_CREATED)
async def create_report(
    category: str = Form(...),
    lat: float = Form(...),
    lng: float = Form(...),
    title: str | None = Form(None),
    description: str | None = Form(None),
    priority: str = Form("Medium"),
    address: str | None = Form(None),
    reported_by: str = Form("Anonymous"),
    image: UploadFile | None = File(None),
) -> dict:
    image_url = None
    if image and image.filename:
        image_url = upload_report_image(image.file, image.filename, image.content_type or "image/jpeg")

    payload = {
        "title": title or f"{category} Issue",
        "description": description,
        "category": category,
        "priority": priority,
        "lat": lat,
        "lng": lng,
        "address": address,
        "image_url": image_url,
        "reported_by": reported_by,
        "status": "Open",
    }

    supabase = get_supabase()
    response = supabase.table("reports").insert(payload).execute()
    if not response.data:
        raise HTTPException(status_code=500, detail="Failed to create report")
    return response.data[0]


@router.patch("/{report_id}/status", response_model=ReportResponse)
async def update_report_status(
    report_id: UUID,
    body: StatusUpdateRequest,
    admin: dict = Depends(get_current_admin),
) -> dict:
    admin_email = admin["email"]
    token = admin["token"]
    payload: dict = {"status": body.status}

    if body.status == "Resolved":
        payload["resolved_by"] = admin_email
        payload["resolved_at"] = datetime.now(timezone.utc).isoformat()
    elif body.previous_status == "Resolved" and body.status != "Resolved":
        payload["resolved_by"] = None
        payload["resolved_at"] = None

    if body.admin_note:
        payload["admin_notes"] = body.admin_note

    supabase = get_supabase_client(token)
    response = (
        supabase.table("reports")
        .update(payload)
        .eq("id", str(report_id))
        .select("*")
        .execute()
    )

    if not response.data:
        raise HTTPException(status_code=404, detail="Report not found")

    try:
        supabase.table("status_history").insert(
            {
                "report_id": str(report_id),
                "from_status": body.previous_status or None,
                "to_status": body.status,
                "changed_by": admin_email,
                "note": body.admin_note or None,
            }
        ).execute()
    except Exception:
        pass

    return response.data[0]


@router.get("/{report_id}/history", response_model=list[StatusHistoryEntry])
async def get_status_history(report_id: UUID) -> list[dict]:
    supabase = get_supabase()
    response = (
        supabase.table("status_history")
        .select("*")
        .eq("report_id", str(report_id))
        .order("changed_at")
        .execute()
    )
    return response.data or []

File: backend/src/routes/reports.js
import { Router } from 'express'
import multer from 'multer'
import { getSupabase, getSupabaseClient } from '../db.js'
import { requireAdmin } from '../middleware/auth.js'
import { uploadReportImage } from '../services/storage.js'

const router = Router()
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 10 * 1024 * 1024 }, // 10MB limit
})

/**
 * GET /api/reports - List all reports ordered by created_at desc
 */
router.get('/', async (req, res, next) => {
  try {
    const supabase = getSupabase()
    const { data, error } = await supabase
      .from('reports')
      .select('*')
      .order('created_at', { ascending: false })

    if (error) {
      return res.status(500).json({ detail: error.message })
    }

    return res.json(data || [])
  } catch (err) {
    next(err)
  }
})

/**
 * POST /api/reports - Create a new report (multipart/form-data)
 */
router.post('/', upload.single('image'), async (req, res, next) => {
  try {
    const { category, lat, lng, title, description, priority = 'Medium', address, reported_by = 'Anonymous' } = req.body

    if (!category || lat === undefined || lng === undefined) {
      return res.status(422).json({ detail: 'category, lat, and lng are required fields' })
    }

    let imageUrl = null
    if (req.file) {
      try {
        imageUrl = await uploadReportImage(req.file.buffer, req.file.originalname, req.file.mimetype)
      } catch (uploadErr) {
        console.error('Image upload failed:', uploadErr)
      }
    }

    const payload = {
      title: title || `${category} Issue`,
      description: description || null,
      category,
      priority: priority || 'Medium',
      lat: parseFloat(lat),
      lng: parseFloat(lng),
      address: address || null,
      image_url: imageUrl,
      reported_by: reported_by || 'Anonymous',
      status: 'Open',
    }

    const supabase = getSupabase()
    const { data, error } = await supabase.from('reports').insert(payload).select('*')

    if (error || !data || data.length === 0) {
      return res.status(500).json({ detail: error?.message || 'Failed to create report' })
    }

    return res.status(201).json(data[0])
  } catch (err) {
    next(err)
  }
})

/**
 * PATCH /api/reports/:id/status - Update report status (admin only)
 */
router.patch('/:id/status', requireAdmin, async (req, res, next) => {
  try {
    const { id } = req.params
    const { status, admin_note = '', previous_status = '' } = req.body

    if (!status) {
      return res.status(400).json({ detail: 'status is required' })
    }

    const adminEmail = req.admin.email
    const token = req.admin.token

    const payload = { status }

    if (status === 'Resolved') {
      payload.resolved_by = adminEmail
      payload.resolved_at = new Date().toISOString()
    } else if (previous_status === 'Resolved' && status !== 'Resolved') {
      payload.resolved_by = null
      payload.resolved_at = null
    }

    if (admin_note) {
      payload.admin_notes = admin_note
    }

    const supabase = getSupabaseClient(token)
    const { data, error } = await supabase
      .from('reports')
      .update(payload)
      .eq('id', id)
      .select('*')

    if (error || !data || data.length === 0) {
      return res.status(404).json({ detail: 'Report not found' })
    }

    // Record status change history
    try {
      await supabase.from('status_history').insert({
        report_id: id,
        from_status: previous_status || null,
        to_status: status,
        changed_by: adminEmail,
        note: admin_note || null,
      })
    } catch (historyErr) {
      console.warn('Failed to insert status history:', historyErr)
    }

    return res.json(data[0])
  } catch (err) {
    next(err)
  }
})

/**
 * GET /api/reports/:id/history - Get status audit trail for a report
 */
router.get('/:id/history', async (req, res, next) => {
  try {
    const { id } = req.params
    const supabase = getSupabase()
    const { data, error } = await supabase
      .from('status_history')
      .select('*')
      .eq('report_id', id)
      .order('changed_at', { ascending: true })

    if (error) {
      return res.status(500).json({ detail: error.message })
    }

    return res.json(data || [])
  } catch (err) {
    next(err)
  }
})

export default router
